pkg/acquisition/modules/syslog/internal/parser/rfc5424/parse.go

```diff
@@ -0,0 +1,398 @@
+package rfc5424
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/crowdsecurity/crowdsec/pkg/acquisition/modules/syslog/internal/parser/utils"
+)
+
+type RFC5424Option func(*RFC5424)
+
+type RFC5424 struct {
+	PRI       int
+	Timestamp time.Time
+	Hostname  string
+	Tag       string
+	Message   string
+	PID       string
+	MsgID     string
+	//
+	len            int
+	position       int
+	buf            []byte
+	useCurrentYear bool //If no year is specified in the timestamp, use the current year
+	strictHostname bool //If the hostname contains invalid characters or is not an IP, return an error
+}
+
+const PRI_MAX_LEN = 3
+
+const NIL_VALUE = '-'
+
+var VALID_TIMESTAMPS = []string{
+	time.RFC3339,
+}
+
+const VALID_TIMESTAMP = time.RFC3339Nano
+
+func WithCurrentYear() RFC5424Option {
+	return func(r *RFC5424) {
+		r.useCurrentYear = true
+	}
+}
+
+func WithStrictHostname() RFC5424Option {
+	return func(r *RFC5424) {
+		r.strictHostname = true
+	}
+}
+
+func (r *RFC5424) parsePRI() error {
+
+	pri := 0
+
+	if r.buf[r.position] != '<' {
+		return fmt.Errorf("PRI must start with '<'")
+	}
+
+	r.position++
+
+	for r.position < r.len {
+		c := r.buf[r.position]
+		if c == '>' {
+			r.position++
+			break
+		}
+		if c < '0' || c > '9' {
+			return fmt.Errorf("PRI must be a number")
+		}
+		pri = pri*10 + int(c-'0')
+		r.position++
+	}
+
+	if pri > 999 {
+		return fmt.Errorf("PRI must be up to 3 characters long")
+	}
+
+	if r.position == r.len && r.buf[r.position-1] != '>' {
+		return fmt.Errorf("PRI must end with '>'")
+	}
+
+	r.PRI = pri
+	return nil
+}
+
+func (r *RFC5424) parseVersion() error {
+	if r.buf[r.position] != '1' {
+		return fmt.Errorf("version must be 1")
+	}
+	r.position += 2
+	if r.position >= r.len {
+		return fmt.Errorf("version must be followed by a space")
+	}
+	return nil
+}
+
+func (r *RFC5424) parseTimestamp() error {
+
+	timestamp := []byte{}
+
+	if r.buf[r.position] == NIL_VALUE {
+		r.Timestamp = time.Now().UTC().Round(0)
+		r.position += 2
+		return nil
+	}
+
+	for r.position < r.len {
+		c := r.buf[r.position]
+		if c == ' ' {
+			break
+		}
+		timestamp = append(timestamp, c)
+		r.position++
+	}
+
+	if len(timestamp) == 0 {
+		return fmt.Errorf("timestamp is empty")
+	}
+
+	if r.position == r.len {
+		return fmt.Errorf("EOL after timestamp")
+	}
+
+	date, err := time.Parse(VALID_TIMESTAMP, string(timestamp))
```

Comment:
from https://www.rfc-editor.org/rfc/rfc5424.html#section-6.2.3 and https://www.rfc-editor.org/rfc/rfc3339 I'm not sure that RFC3339Nano fits all the use cases ? Didn't look at the tests yet, but should at least be :

`1985-04-12T23:20:50.52Z`
`2003-10-11T22:14:15.003Z`
`1985-04-12T19:20:50.52-04:00`
`2003-08-24T05:14:15.000003-07:00`

Reply:
RFC3339Nano handles all those timestamps.